Bluetooth: Host: Use system work queue instead of BT RX

LingaoM opened the pull request:
When `CONFIG_BT_RECV_IS_RX_THREAD` not selected, for example, under `native` or `rpmsg` configuration.

Instead of recreating a `BT RX`, use the system work queue directly.

This PR tests almost the projects included in `samples/bluetooth/*` and has been tested by `autopts` under `native_posix` configuration.

The default system work queue can not be automatically configured, and can not pollute the system configuration. 
Therefore, it is proposed to redefine the operation in the Bluetooth subsystem to realize automatic configuration.

Also create a discussion https://github.com/zephyrproject-rtos/zephyr/discussions/43509 about whether there is a better way to automate configuration and get everyone to agree. :)

Signed-off-by: Lingao Meng <menglingao@xiaomi.com>

Thalley commented:
> This PR will replace `rx_thread` with system work thread.
> 
> To reduce the overhead of system resources.
> 
> Signed-off-by: Lingao Meng [menglingao@xiaomi.com](mailto:menglingao@xiaomi.com)

Do you have any numbers on how much this saves us? Are there any performance degradation to consider, or other side effects?

LingaoM commented:
> Do you have any numbers on how much this saves us?

At least one stack released. https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/host/Kconfig#L83

LingaoM commented:
> Are there any performance degradation to consider

Current there no way to test performance accurate data.

Thalley commented:
> > Do you have any numbers on how much this saves us?
> 
> At least one stack released. https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/host/Kconfig#L83

Yes, but moving from a the thread stack to the workqueue stack doesn't really save much if `SYSTEM_WORKQUEUE_STACK_SIZE` needs to be similar increased to support the same scenarios. 

I am not sure how much RAM is required for the k_threads, but if they are anything like a POSIX thread then it's about ~300 bytes, whereas the workqueue (`struct k_work`) is about ~50, so that's a ~250 bytes saving it seems, which isn't too bad.

LingaoM commented:
> > > Do you have any numbers on how much this saves us?
> > 
> > 
> > At least one stack released. https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/host/Kconfig#L83
> 
> Yes, but moving from a the thread stack to the workqueue stack doesn't really save much if `SYSTEM_WORKQUEUE_STACK_SIZE` needs to be similar increased to support the same scenarios.
> 
> I am not sure how much RAM is required for the k_threads, but if they are anything like a POSIX thread then it's about ~300 bytes, whereas the workqueue (`struct k_work`) is about ~50, so that's a ~250 bytes saving it seems, which isn't too bad.

Bye the way, `rx_thread_data` = 128 Bytes.

Thalley commented:
> By the way, `rx_thread_data` = 128 Bytes.

So the savings are more like 78 bytes of RAM when not accounting for the potentially smaller stack size.

LingaoM commented:
> > By the way, `rx_thread_data` = 128 Bytes.
> 
> So the savings are more like 78 bytes of RAM when not accounting for the potentially smaller stack size.

Not really, `rx_thread_data` = 128 Bytes, `rx_thread_stack` = 1024 by default.

So, can save at least 1KB RAM.

Thalley commented:
> > > By the way, `rx_thread_data` = 128 Bytes.
> > 
> > 
> > So the savings are more like 78 bytes of RAM when not accounting for the potentially smaller stack size.
> 
> Not really, `rx_thread_data` = 128 Bytes, `rx_thread_stack` = 1024 by default.
> 
> So, can save at least 1KB RAM.

Assuming that `SYSTEM_WORKQUEUE_STACK_SIZE` doesn't need to be increased to handle the additional load.

LingaoM commented:
> > > > By the way, `rx_thread_data` = 128 Bytes.
> > > 
> > > 
> > > So the savings are more like 78 bytes of RAM when not accounting for the potentially smaller stack size.
> > 
> > 
> > Not really, `rx_thread_data` = 128 Bytes, `rx_thread_stack` = 1024 by default.
> > So, can save at least 1KB RAM.
> 
> Assuming that `SYSTEM_WORKQUEUE_STACK_SIZE` doesn't need to be increased to handle the additional load.

I don't think it will increase a lot, depending on the application scenario. Of course, for complex scenarios, we still need to adjust the stack space of Rx, so when we remove Rx, it's not bad.

At least `min (syswork stack size, RX stack size) `can be saved.

Thalley commented:
> > > > > By the way, `rx_thread_data` = 128 Bytes.
> > > > 
> > > > 
> > > > So the savings are more like 78 bytes of RAM when not accounting for the potentially smaller stack size.
> > > 
> > > 
> > > Not really, `rx_thread_data` = 128 Bytes, `rx_thread_stack` = 1024 by default.
> > > So, can save at least 1KB RAM.
> > 
> > 
> > Assuming that `SYSTEM_WORKQUEUE_STACK_SIZE` doesn't need to be increased to handle the additional load.
> 
> I don't think it will increase a lot, depending on the application scenario. Of course, for complex scenarios, we still need to adjust the stack space of Rx, so when we remove Rx, it's not bad.
> 
> At least `min (syswork stack size, RX stack size) ` bytes of ram can be saved.

How do we ensure that this change doesn't break any of the Bluetooth Samples? The babblesim tests will run in CI, but the samples won't. Perhaps we need to run a few select BT samples to ensure that this change doesn't affect them in terms of whether the `SYSTEM_WORKQUEUE_STACK_SIZE` is adequate for those.

thoh-ot commented:
With this change, doesn't it introduce the possibility of blocking the system work queue? If a Bluetooth host callback calls back into the host and it cannot allocate a GATT TX buffer, the system work thread is now blocked.

Thalley commented:
> With this change, doesn't it introduce the possibility of blocking the system work queue? If a Bluetooth host callback calls back into the host and it cannot allocate a GATT TX buffer, the system work thread is now blocked.

That is still possible today if someone uses the system workqueue with a timer to send GATT data, isn't it?

LingaoM commented:
> How do we ensure that this change doesn't break any of the Bluetooth Samples? The babblesim tests will run in CI, but the samples won't. Perhaps we need to run a few select BT samples to ensure that this change doesn't affect them in terms of whether the `SYSTEM_WORKQUEUE_STACK_SIZE` is adequate for those.

Yes, i'm test a few samples(peripheral, central, scan_adv, mesh), but not coverage all.

thoh-ot commented:
> > With this change, doesn't it introduce the possibility of blocking the system work queue? If a Bluetooth host callback calls back into the host and it cannot allocate a GATT TX buffer, the system work thread is now blocked.
> 
> That is still possible today if someone uses the system workqueue with a timer to send GATT data, isn't it?

Correct; my point was that now we are adding another one, just from a "normal" host callback

LingaoM commented:
ref: https://github.com/zephyrproject-rtos/zephyr/discussions/40243

alwa-nordic commented:
What do we need to do to increase our confidence in the correctness of this change? Can we test this with some stress tests?

Thalley commented:
> What do we need to do to increase our confidence in the correctness of this change? Can we test this with some stress tests?

We already have quite a few LE Audio GATT based babblesim tests that will require and send quite a lot of data in a short time span. 

We are also adding L2CAP ecred bsim tests, as well as generic GATT bsim test (https://github.com/zephyrproject-rtos/zephyr/pull/41630). We could expand those to test something like this specifically.

LingaoM commented:
Any problem?

LingaoM commented:
https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/controller/Kconfig#L104

Auto selected this config, therefore, this task does not exist by default.

Thalley commented:
Besides removing the overhead of having this thread, we also change 
1) The priority of the "RX thread" as the workqueue thread has a lower (by default) priority than the current RX thread
2) The thread that callbacks are called from will some unified, as some is currently being called from the RX thread, and some from the workqueue thread.

LingaoM commented:
> Besides removing the overhead of having this thread, we also change 
> 
> 1) The priority of the "RX thread" as the workqueue thread has a lower (by default) priority than the current RX thread

I have no idea about that.

> 2) The thread that callbacks are called from will some unified, as some is currently being called from the RX thread, and some from the workqueue thread. 

Sounds like the second is an advantage

Thalley commented:
> > 2. The thread that callbacks are called from will some unified, as some is currently being called from the RX thread, and some from the workqueue thread.
> 
> Sounds like the second is an advantage

I absolutely agree :) But it may impact how some applications work if they use this specific knowledge for something.

LingaoM commented:
> > > 2. The thread that callbacks are called from will some unified, as some is currently being called from the RX thread, and some from the workqueue thread.
> 
> > 
> 
> > Sounds like the second is an advantage
> 
> 
> 
> I absolutely agree :) But it may impact how some applications work if they use this specific knowledge for something. 
> 
> 

According https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/controller/Kconfig#L104 In fact, the "BT RX" is not used by default. I wonder when this task will be used. It seems not?

Thalley commented:
> According https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/controller/Kconfig#L104 In fact, the "BT RX" is not used by default. I wonder when this task will be used. It seems not?

AFAIK, it is used for host builds, e.g. when building for `native_posix` or `nrf5340dk_nrf5340_cpuapp`, but not 100% sure on that.

LingaoM commented:
> > According https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/controller/Kconfig#L104 In fact, the "BT RX" is not used by default. I wonder when this task will be used. It seems not?
> 
> AFAIK, it is used for host builds, e.g. when building for `native_posix` or `nrf5340dk_nrf5340_cpuapp`, but not 100% sure on that.


https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/bluetooth/hci/Kconfig#L14-L20 also select this.

Thalley commented:
@LingaoM Have this been tested extensively when `CONFIG_BT_RECV_IS_RX_THREAD=y`?

LingaoM commented:
The default is `y`, so the default has been widely tested.

joerchan commented:
> The default is `y`, so the default has been widely tested.

The "default" depends on the HCI driver being used. Combined build with zephyr controller will have this selected. Other controllers, or non-combined builds will not have always have this selected.

LingaoM commented:
> CONFIG_BT_RECV_IS_RX_THREAD

I‘m build native_posix with bluetooth samples, include peripheral, central, mesh.

LingaoM commented:
For other with `iso` or `le audio` testing, may @Thalley more appropriate?

LingaoM commented:
> In main we see a delay of about 0.4ms for 10000 transmissions in main, and 1.2ms in this PR. In terms of ISO this is not insignificant.

I think the savings of more than 2KB of RAM are worth the 0.008% overhead.

Thalley commented:
> > In main we see a delay of about 0.4ms for 10000 transmissions in main, and 1.2ms in this PR. In terms of ISO this is not insignificant.
> 
> I think the savings of more than 2KB of RAM are worth the 0.008% overhead.

I'd need to talk to the people actively using audio over ISO to see what they think. It's not a 0.008% overhead in my opinion, but rather a 200% overhead (going from 0.4ms to 1.2ms). 

I'll also run some more tests on connected ISO benchmarks (where there is far more traffic given that it is bidirection), and see what the effect of this is there.

Thalley commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -3387,6 +3387,17 @@ void hci_event_prio(struct net_buf *buf)
+#if !defined(CONFIG_BT_RECV_IS_RX_THREAD)
+static void rx_queue_put(struct net_buf *buf)
+{
+	net_buf_put(&bt_dev.rx_queue, buf);
+
+	if (!k_work_is_pending(&rx_work)) {
+		k_work_submit(&rx_work);
+	}
```
We should check the return code of k_work_submit and log errors. 

The check for `k_work_is_pending` is not needed, as `k_work_submit` should handle that itself:

```c
 * @retval 0 if work was already submitted to a queue
 * @retval 1 if work was not submitted and has been queued to @p queue
 * @retval 2 if work was running and has been queued to the queue that was
 * running it
```

```suggestion
	int err;
	
	net_buf_put(&bt_dev.rx_queue, buf);

	err = k_work_submit(&rx_work); 
	if (err < 0) {
		BT_ERR("Could not submit rx_work: %d", err);
	}
```

Thalley commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -3566,12 +3576,13 @@ static void hci_rx_thread(void)
+	 * gets empty.
+	 */
+	if (!k_fifo_is_empty(&bt_dev.rx_queue)) {
+		k_work_submit(&rx_work);
```
Please check return code and log errors

Thalley commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -3387,6 +3387,20 @@ void hci_event_prio(struct net_buf *buf)
+	net_buf_put(&bt_dev.rx_queue, buf);
+
+	err = k_work_submit(&rx_work);
+	if (err) {
```
```suggestion
	if (err < 0) {
```

k_work_submit will return 0, 1 or 2 on success :)

Thalley commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -3566,12 +3580,18 @@ static void hci_rx_thread(void)
+	}
+
+	err = k_work_submit(&rx_work);
+	if (err) {
```
```suggestion
	if (err < 0) {
```

Thalley reviewed (approved):
There may be some unforeseen consequences that haven't been tested, but I have no experience any issues with this yet. 

The main change is that BT RX priority will be lowered, but it's unsure if that that is a good or bad thing. 

A clear advantage of this is that all async callbacks from the BT subsystem to the application will now be from the same thread, instead of a mix between the RX thread and the workqueue thread. 

Suggest to approve this, and if it causes any issues we can revert it.

haoyanbiao reviewed (dismissed):
LGTM

jhedberg commented:
Should the default values for `BT_RX_STACK_SIZE` in `subsys/bluetooth/host/Kconfig` be fine-tuned together with this change?

LingaoM commented:
> Should the default values for `BT_RX_STACK_SIZE` in `subsys/bluetooth/host/Kconfig` be fine-tuned together with this change?

`BT_RX_STACK_SIZE` will used by controller, i don't see anything that needs to be changed.

jhedberg commented:
Actually, there is one possible change to Kconfig that you might want to do: add this to the definition of `BT_RX_STACK_SIZE`: `depends on !BT_RECV_IS_RX_THREAD`. Makes sense?

LingaoM commented:
> Actually, there is one possible change to Kconfig that you might want to do: add this to the definition of `BT_RX_STACK_SIZE`: `depends on !BT_RECV_IS_RX_THREAD`. Makes sense?
> 
> 
> 
> By making that change the user gets a warning/error if they try to set `BT_RX_STACK_SIZE` for a configuration where the value will not be used.

https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/controller/Kconfig#L104

will select this， so if `BT_RX_STACK_SIZE`: `depends on !BT_RECV_IS_RX_THREAD` will be error.

jhedberg commented:
@LingaoM why would there be an error? As far as I can see the error would only happen if the app tried to set `BT_RX_STACK_SIZE` in its `prj.conf` when using LL_SW.

jhedberg commented:
Correction, with the LL_SW case you *could* set BT_RX_STACK_SIZE

jhedberg commented:
@LingaoM maybe you missed the `!` in my proposal?

jhedberg commented:
@LingaoM yeah, you're right sorry for the confusion - I guess it should be exactly the other way around, i.e. *without* the `!`: `depends on BT_RECV_IS_RX_THREAD`, right?

Thalley commented on subsys/bluetooth/host/Kconfig:
```diff
@@ -82,7 +82,7 @@ config BT_RECV_IS_RX_THREAD
+	depends on BT_RECV_IS_RX_THREAD
 	default 512 if BT_HCI_RAW
 	default 2048 if BT_MESH
 	default 2200 if BT_SETTINGS
```
Shouldn't we add/move these defaults for `BT_MESH` and `BT_SETTINGS` to `SYSTEM_WORKQUEUE_STACK_SIZE`? 

Have this change been tested with BT_SETTINGS enabled or with MESH?

LingaoM replied:
First of all, this `BT_RX_STACK_SIZE` still needs to be used by the soft based controller, so we can't remove it.

LingaoM replied:
Added.

nashif commented on kernel/Kconfig:
```diff
@@ -422,6 +422,8 @@ menu "Work Queue Options"
 config SYSTEM_WORKQUEUE_STACK_SIZE
 	int "System workqueue stack size"
 	default 4096 if COVERAGE
+	default 2048 if BT_MESH && !BT_RECV_IS_RX_THREAD
```
let's not pollute this kernel configuration with subsystem related settings.

LingaoM replied:
Removed.

Thalley replied:
Removing this will likely cause issues for some Mesh cases. 

Since BT will now use `SYSTEM_WORKQUEUE_STACK_SIZE` and/or applications using `BT_SETTINGS`, as well as `BT_AUDIO`. 

How do ensure (or at least indicate) that specific BT configurations need increased `SYSTEM_WORKQUEUE_STACK_SIZE` values?

jhedberg replied:
If we’re not allowed to touch Kconfig entries outside of the subsystem like @nashif says (which does make sense) I think the other option is to use build asserts?

Thalley replied:
> If we’re not allowed to touch Kconfig entries outside of the subsystem like @nashif says (which does make sense) I think the other option is to use build asserts?

If we go from a suggested default (like here) to a specific and hard requirement like BUILD_ASSERT, then we need to spend some additional time on calculating the stack usage, as we probably shouldn't require e.g. 2048 octets for the stack if 1200 is sufficient. 

@Casper-Bonde-Bose You mentioned that native_posix stack sizes were not affected by this config. Does that apply to babblesim as well (ping @aescolar )? In other words, does this value affect the babblesim test cases?

LingaoM replied:
@Thalley babblesim use `Controller+Host`, which `CONFIG_BT_RECV_IS_RX_THREAD` selected by https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/controller/Kconfig#L104, so this change not affect babblesim test cases.

Thalley commented on subsys/bluetooth/host/CMakeLists.txt:
```diff
@@ -126,3 +126,9 @@ if(CONFIG_BT_CONN_DISABLE_SECURITY)
   Do not use in production."
     )
 endif()
+if((NOT CONFIG_BT_RECV_IS_RX_THREAD) AND
+   (CONFIG_SYSTEM_WORKQUEUE_STACK_SIZE LESS 2200))
+  message(WARNING "For Bluetooth Settings function.
+  It is recommended that the system work thread stack space is 2200 and above."
+    )
+endif()
```
Seems to be missing a check for CONFIG_BT_SETTINGS, but otherwise this is a decent workaround if we can't add the dependency for `CONFIG_SYSTEM_WORKQUEUE_STACK_SIZE`

LingaoM replied:
Good eyesight.

LingaoM commented:
@Thalley @jhedberg @nashif CC

jhedberg commented:
Should we at the same time update `prj.conf` files of in-tree sample/test applications, to make sure that they adhere to the minimum system workqueue stack sizes?

LingaoM commented:
> Should we at the same time update `prj.conf` files of in-tree sample/test applications, to make sure that they adhere to the minimum system workqueue stack sizes?

There are currently no relevant tests related to this configuration, at least I haven't found.

jhedberg commented:
It still bugs me that we can't automatically influence sane default now for a thread stack that the Bluetooth subsystem uses. `prj.conf` is for an application to state its needs. Here however we're talking about a Zephyr subsystem that's using a shared resource (the system workqueue) but can't directly (and in an automated way) express its requirements for that resource. 

@nashif I agree that it's not good/clean to have a kernel Kconfig option have conditional characteristics depending on non-kernel subsystems (like Bluetooth), but having to use an application feature (`prj.conf`) to encode a subsystem's requirements isn't ideal either, so might be good to brainstorm if some third option might exist (or be created).

Thalley reviewed (changes requested):
Found a reference to the RX thread:

```c
 * The Notification callback @p params->notify is also called from the BT RX
 *  thread.
 ```

Such references should probably also be updated. Please search and check if there is more.

Thalley commented:
> If we used combination with controller, that's actually called from BT RX. So we change to what?

Good question. This change basically changes the thread of callbacks based on other configurations. I think it would make sense to somehow document that. 

We could remove all references to specific threads, and then describe which thread is used for callbacks in a central place. 
Alternatively, we should define, for each callback, which thread calls it (and in which configuration).

carlescufi commented:
This is an interesting proposal, but I am afraid it will introduce deadlocks. I really do not know how someone can check or prove that there will be no additional deadlocks, but we need at least an analysis before we consider merging this.

Thalley commented:
> > This is an interesting proposal, but I am afraid it will introduce deadlocks. I really do not know how someone can check or prove that there will be no additional deadlocks, but we need at least an analysis before we consider merging this.
> 
> I have tested all mostly Bluetooth samples with @Thalley , and bbsim seems to be normal.
> 
> [#40273 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/40273#issuecomment-1038988152)

Perhaps we can also run it through auto-pts to gather additional evidence

LingaoM commented:
> > > This is an interesting proposal, but I am afraid it will introduce deadlocks. I really do not know how someone can check or prove that there will be no additional deadlocks, but we need at least an analysis before we consider merging this.
> > 
> > 
> > I have tested all mostly Bluetooth samples with @Thalley , and bbsim seems to be normal.
> > [#40273 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/40273#issuecomment-1038988152)
> 
> Perhaps we can also run it through auto-pts to gather additional evidence

I will use the `native_posix` build https://github.com/intel/auto-pts#running-in-clientserver-mode with autopts and synchronize later.

LingaoM commented:
I'll use this branch to compile native_posix tests. Here are some test reports:

![9707da3a6edaa51b17b8444fcc00c50](https://user-images.githubusercontent.com/26378606/155680262-590c20f2-5941-4d4f-8d58-49c7f41c06b9.png)
![a6dfd3dbbece169a7fb30c0b164e9ce](https://user-images.githubusercontent.com/26378606/155680272-0987ff5e-2f6a-405f-babf-f809fbe04616.png)
![fc295a5b9a00cecd28cb073f9cbffef](https://user-images.githubusercontent.com/26378606/155680279-39aefc68-3de9-4c2a-92b9-47a43eb41359.png)
![ebb2af1679c20078c24e0da703cf940](https://user-images.githubusercontent.com/26378606/155680289-f5d3d12c-6d61-463a-b288-be60c2020995.png)
![5244fbb9c7f030a09f5005213e88e4e](https://user-images.githubusercontent.com/26378606/155680295-c29f9799-e249-4d23-84ac-abb6a074d1e0.png)

- GATT/SR/UNS/BI-01-C
- MESH/NODE/TNPT/BV-04-C
- MESH/NODE/TNPT/BV-05-C
- MESH/NODE/TNPT/BI-01-C

These failed test items have the **same results** on the `main` branch. Therefore, I do not intend to discuss another problem here, but it can be explained that the modification of PR does not introduce new problems.

Thalley commented on include/bluetooth/gatt.h:
```diff
@@ -1217,10 +1217,12 @@ struct bt_gatt_exchange_params {
  *
  *  This function will block while the ATT request queue is full, except when
- *  called from the BT RX thread, as this would cause a deadlock.
+ *  called from the BT RX thread or System work queue thread when
+ *  @kconfig{CONFIG_BT_RECV_IS_RX_THREAD} not selected, as this would cause a deadlock.
```
I wonder if we could make it something simpler like 

```suggestion
 *  called from a Bluetooth callback, as this would cause a deadlock.
```

and still keep the same amount of information?

Thalley commented on tests/bluetooth/tester/overlay-native.conf:
```diff
@@ -1,3 +1,4 @@
 CONFIG_UART_PIPE=n
 CONFIG_SERIAL=y
 CONFIG_UART_NATIVE_POSIX=y
+CONFIG_SYSTEM_WORKQUEUE_STACK_SIZE=2048
```
This should probably be done for all existing LE Audio, Mesh and samples that enables BT_SETTINGS as well with their respective values of course, as to not change anything in those.

LingaoM replied:
Need to change here?

Thalley replied:
No this is likely fine, but there's numerous LE Audio and Mesh samples that now may have insufficient stack sizes, as they went from 2048 to 1024, as well as the samples that used BT_SETTINGS that should have `CONFIG_SYSTEM_WORKQUEUE_STACK_SIZE=2200`

Thalley replied:
In other words, no samples should compile with the newly added warnings.

Thalley replied:
@LingaoM Will you update the configs of existing samples that probably expected this stack size?

LingaoM replied:
I tried to search for other samples and found that other samples do not take effect under this condition（BT_RECV_IS_RX_THREAD=n）, including all bbsim.

As you say, add CMake warning for insufficient thread stack size can also provide way to let's user known it's should increase stack size when change default configuration.

Thalley replied:
> I tried to search for other samples and found that other samples do not take effect under this condition（BT_RECV_IS_RX_THREAD=n）, including all bbsim.

That depends on the board being compiled though, e.g. when using the nRF53 with split architecture. 

> As you say, add CMake warning for insufficient thread stack size can also provide way to let's user known it's should increase stack size when change default configuration.

Exactly, and now the samples that enables BT_AUDIO, BT_MESH or BT_SETTINGS will get warnings when compiling for certain boards.

LingaoM replied:
Remove this line, since find already in https://github.com/zephyrproject-rtos/zephyr/blob/main/tests/bluetooth/tester/prj.conf#L3

LingaoM commented:
> * GATT/SR/UNS/BI-01-C
> * MESH/NODE/PROV/BV-12-C (PASSED 2nd)
> * MESH/NODE/TNPT/BV-04-C
> * MESH/NODE/TNPT/BV-05-C
> * MESH/NODE/TNPT/BI-01-C

I‘m use 8.0.3 testing this failed case with this PR again, then PASSED.

BTW: Current PTS Version 8.2.0

nashif commented:
> @nashif I agree that it's not good/clean to have a kernel Kconfig option have conditional characteristics (e.g. range or default value) depending on non-kernel subsystems (like Bluetooth), but having to use an application feature (`prj.conf`) to encode a subsystem's requirements isn't ideal either, so might be good to brainstorm if some third option might exist (or be created).

what if you have to subsystems doing this and requesting two different values of a system resource?

jhedberg commented:
> what if you have to subsystems doing this and requesting two different values of a system resource?

With the way Kconfig is now implemented the result would depend on how the defaults are ordered. That's also why I suggested to brainstorm of some third option that would e.g. allow subsystems to declare a minimum requirement and then the maximum of those would get selected at build time.

andyross commented:
> option that would e.g. allow subsystems to declare a minimum requirement

You can get a long way with just

    /* This subsystem has nontrivial stack requirements, make sure there is enough */
    #define MIN_STACK  .... /* <--- almost certainly platform-dependent */
    BUILD_ASSERT(CONFIG_SYSTEM_WORKQUEUE_STACK_SIZE > MIN_STACK);

Note that stack sizes are **very** sensitive to architecture and kernel configuration choices elsewhere, so a simple default in a kconfig file isn't likely to have much value.  It'll work for the platforms it's tested on, and then blow up when someone turns on canaries or analysis, or sets CONFIG_MISRA_SANE=y (forces off the use of alloca() and uses big stack buffers), or runs on Xtensa or SPARC (known stack hogs), etc...

Thalley commented:
> > option that would e.g. allow subsystems to declare a minimum requirement
> 
> You can get a long way with just
> 
> ```
> /* This subsystem has nontrivial stack requirements, make sure there is enough */
> #define MIN_STACK  .... /* <--- almost certainly platform-dependent */
> BUILD_ASSERT(CONFIG_SYSTEM_WORKQUEUE_STACK_SIZE > MIN_STACK);
> ```
> 
> Note that stack sizes are **very** sensitive to architecture and kernel configuration choices elsewhere, so a simple default in a kconfig file isn't likely to have much value. It'll work for the platforms it's tested on, and then blow up when someone turns on canaries or analysis, or sets CONFIG_MISRA_SANE=y (forces off the use of alloca() and uses big stack buffers), or runs on Xtensa or SPARC (known stack hogs), etc...

Not to mention application code that may reside in the callbacks called in the system workqueue thread (was recently an issue for me). 

Defining a strict minimum (using e.g. BUILD_ASSERT) may be different than we want, if we just want a safe default stack value. The previous solution was to rely on `default` which can of course be overridden (larger or smaller) by the application, and thus leaving optimization to the application. If we use a BUILD_ASSERT (and any other strict method), then the application will not be able to reduce the stack to an optimized value. 

For these reasons, I think the solution that @LingaoM have provided using a CMake warning is a better solution, as we can "get away with" suggesting larger stack sizes than the absolute minimum.

LingaoM commented:
The Kconfig we are discussing here, I think it has been separated from the PR itself. Considering that this PR has been staying for a long time, I suggest that we can continue to discuss whether this PR can be merge into the community. If there is a suitable solution to the Kconfig issue later, I'd be happy to resubmit.

LingaoM commented on samples/bluetooth/unicast_audio_server/boards/nrf5340dk_nrf5340_cpuapp.conf:
```diff
@@ -1,2 +1,4 @@
+CONFIG_SYSTEM_WORKQUEUE_STACK_SIZE=2048
```
@Thalley For a specific configuration, and the defconfig of sysworkq is not set, only this one is found.

Thalley replied:
Hmm, since this change will only affect a subset of configuration (e.g. whenever nRF5340 is used), perhaps it's not worth going through all samples and add this (even though we should to retain the existing functionality). Perhaps it is just better to use the warning so that the users of those samples can either set the stack size manually, or add a board configuration.

LingaoM commented on subsys/bluetooth/Kconfig:
```diff
@@ -39,6 +39,16 @@ config BT_CUSTOM
+
 # The Bluetooth subsystem requires the system workqueue to execute at
 # a cooperative priority.
 config SYSTEM_WORKQUEUE_PRIORITY
```
@jhedberg @Thalley @nashif 

>  if some third option might exist (or be created).

Refer to how the priority is used, so that it can be automatically configured without polluting the code in Kernel Kconfig, maybe this is the third way you want. :)

LingaoM commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -3373,13 +3373,33 @@ void hci_event_prio(struct net_buf *buf)
 	}
+#else
+	if (bt_recv_thread_id == NULL) {
+		bt_recv_thread_id = &k_sys_work_q.thread;
```
I think it's a bugfix here, because it wasn't assigned a value before, which would cause bt_recv_thread_id to never be equal to BT RX.

Thalley replied:
From `hci_core.c`:
```c
int bt_recv(struct net_buf *buf)
{
	if (bt_recv_thread_id == NULL) {
		bt_recv_thread_id = k_current_get();
	}
```

LingaoM replied:
I don't think so, according PR(https://github.com/zephyrproject-rtos/zephyr/pull/40056), It doesn't take into account that when this `CONFIG_BT_RECV_IS_RX_THREAD=n`.

LingaoM replied:
bt_recv may be call for `h4.c` or `h5.c` or `RPMsg` etc..

https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/host/att.c#L3240

Will always `false`  when `CONFIG_BT_RECV_IS_RX_THREAD=n`.

alwa-nordic replied:
I'm confused. I believe `CONFIG_BT_RECV_IS_RX_THREAD` does not matter here. `bt_recv_thread_id` is used to determine if `bt_recv()` is on call stack. Are you saying [ATT request allocation] can deadlock if `CONFIG_BT_RECV_IS_RX_THREAD=n`?:

[ATT request allocation]: https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/host/att.c#L3248

LingaoM replied:
yes

LingaoM replied:
The bt_recv_thread_id you added here does not represent BT RX in hci_core.c. If CONFIG_BT_RECV_IS_RX_THREAD=n, ATT Requst Allocation in BT RX will still cause deadlock.

LingaoM replied:
When CONFIG_BT_RECV_IS_RX_THREAD=n, all data will be processed by BT RX through net_buf_put, including ACL, so we should also consider this situation.

alwa-nordic replied:
I see. I'll open a new issue.

LingaoM replied:
Yes, I tried to fix it together with this pr, and in a separate commit, just to confirm with you.

alwa-nordic replied:
Here is the issue: #43448

LingaoM commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -3378,7 +3378,11 @@ k_tid_t bt_recv_thread_id;
 		bt_recv_thread_id = k_current_get();
+#else
+		bt_recv_thread_id = &rx_thread_data;
+#endif
```
@alwa-nordic CC

LingaoM commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -434,6 +437,10 @@ static void hci_acl(struct net_buf *buf)
 
+	if (bt_recv_thread_id == NULL) {
+		bt_recv_thread_id = k_current_get();
+	}
```
@alwa-nordic Please review for this.

LingaoM commented:
@jhedberg @Thalley Let's go on, this PR has been for a long time.

Thalley commented on include/bluetooth/gatt.h:
```diff
@@ -1217,10 +1217,11 @@ struct bt_gatt_exchange_params {
  *
  *  This function will block while the ATT request queue is full, except when
- *  called from the BT RX thread, as this would cause a deadlock.
+ *  called from the Bluetooth callback, as this would cause a deadlock.
```
Applies multiple places:
```suggestion
 *  called from a Bluetooth callback, as this would cause a deadlock.
```
or
```suggestion
 *  called from the Bluetooth callback thread, as this would cause a deadlock.
```

Thalley commented on include/bluetooth/gatt.h:
```diff
@@ -1230,8 +1231,8 @@ struct bt_gatt_exchange_params {
  *  Allow a pending request to resolve before retrying, or call this function
- *  outside the BT RX thread to get blocking behavior. Queue size is controlled
- *  by @kconfig{CONFIG_BT_L2CAP_TX_BUF_COUNT}.
+ *  outside the BT RX thread or System work queue to get blocking behavior.
```
```suggestion
 *  outside a Bluetooth callback to get blocking behavior.
```
or
```suggestion
 *  outside the Bluetooth callback thread to get blocking behavior.
```

Thalley commented on kernel/Kconfig:
```diff
@@ -419,10 +419,14 @@ endif # THREAD_RUNTIME_STATS
 
 menu "Work Queue Options"
+
+config SYSTEM_WORKQUEUE_STACK_SIZE_CUSTOM
+	bool
```
I am not sure I understand this Kconfig, nor its purpose. Can you elaborate?

LingaoM replied:
If other components want to customize the stack space, can select this option.

LingaoM replied:
By default, this option is not well selected, so it does not affect the default configuration well.

Thalley replied:
Can you elaborate a bit more? 

How does other components customize the stack space by setting this option? It seems that it just removes a default value.

LingaoM replied:
Indeed, other components need to redefine this configuration, so that the stack space can be automatically configured without polluting the system code.

LingaoM replied:
As I did in the Bluetooth component.

LingaoM replied:
I have done related tests. In this way, it can be configured automatically and does not need to be explicitly configured in `prj.conf`.

Thalley replied:
> As I did in the Bluetooth component.

Do you mean the default values set in

```
config SYSTEM_WORKQUEUE_STACK_SIZE
	default 3092 if BT_MESH_GATT_CLIENT
	default 2200 if BT_SETTINGS
	default 2048 if BT_MESH
	default 2048 if BT_AUDIO
```
?

Please be aware that these only set default values, and can be configured by the application. This is similar to what we have for the RX thread stack size indeed, but it was pretty clear from https://github.com/zephyrproject-rtos/zephyr/pull/40273#discussion_r811542698 that this was unwanted for a kernel config.

LingaoM replied:
> ?

Yes

> but it was pretty clear from https://github.com/zephyrproject-rtos/zephyr/pull/40273#discussion_r811542698 that this was unwanted for a kernel config.

I think the point here is to prohibit adding code for other components under the system directory, but not prohibiting redefining under other components.

May ask @nashif or @jhedberg Views ?

Thalley commented on subsys/bluetooth/Kconfig:
```diff
@@ -39,6 +39,16 @@ config BT_CUSTOM
 
+if !BT_RECV_IS_RX_THREAD
+
+config SYSTEM_WORKQUEUE_STACK_SIZE
+	default 3092 if BT_MESH_GATT_CLIENT
+	default 2200 if BT_SETTINGS
+	default 2048 if BT_MESH
+	default 2048 if BT_AUDIO
```
I thought we didn't want to do this?

LingaoM replied:
https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/Kconfig#L44
Refer to this usage

LingaoM commented:
Since there is no consensus on automation configuration and no constructive comments from anyone other than @Thalley , considering how long this PR exists, I do not intend to continue to invest in a pointless thing, so choose to close and wait @jhedberg so called 3rd option?

Also for the needs of automated configuration, it is necessary for me to open a separate issure for discussion, not here!

LingaoM commented:
https://github.com/zephyrproject-rtos/zephyr/discussions/43509

LingaoM commented:
> It still bugs me that we can't automatically influence sane defaults for a thread stack that the Bluetooth subsystem uses. prj.conf is for an application to state its needs. 

@jhedberg For the existing solutions, I still hope to hear your opinions！

carlescufi commented:
> Since there is no consensus on automation configuration and no constructive comments from anyone other than @Thalley , considering how long this PR exists, I do not intend to continue to invest in a pointless thing, so choose to close and wait @jhedberg so called 3rd option?

We discussed this today and, in order to be more conservative, how about we change it in this way:

Change `CONFIG_BT_RECV_IS_RX_THREAD` into a `choice`: `CONFIG_BT_RECV_CONTEXT` with the following options (names can be discussed further of course):
- `CONFIG_BT_RECV_DIRECT`
- `CONFIG_BT_RECV_THREAD`
- `CONFIG_BT_RECV_SYS_WORKQ`
- `CONFIG_BT_RECV_USR_WORKQ`

This way users would be able to choose what to run most of the BLE stack on, they wouldn't be forced to a single model.

LingaoM committed:
Bluetooth: Host: Add choice select whether BT RX

Change CONFIG_BT_RECV_IS_RX_THREAD into a
choice:CONFIG_BT_RECV_CONTEXT with the following options
(names can be discussed further of course):

    CONFIG_BT_RECV_BLOCKING
    CONFIG_BT_RECV_WORKQ_BT
    CONFIG_BT_RECV_WORKQ_SYS

This way users would be able to choose what to run most of
the BLE stack on, they wouldn't be forced to a single model.

We would default to CONFIG_BT_RECV_BLOCKING so that we wouldn't
need to change the system workqueue stack size by default, instead
asking users to do so if they select the CONFIG_BT_RECV_WORKQ_SYS option

Signed-off-by: Lingao Meng <menglingao@xiaomi.com>

Thalley commented:
> > Since there is no consensus on automation configuration and no constructive comments from anyone other than @Thalley , considering how long this PR exists, I do not intend to continue to invest in a pointless thing, so choose to close and wait @jhedberg so called 3rd option?
> 
> @LingaoM @jhedberg @Thalley We discussed this today and, in order to be more conservative, how about we change it in this way:
> 
> Change `CONFIG_BT_RECV_IS_RX_THREAD` into a `choice`: `CONFIG_BT_RECV_CONTEXT` with the following options (names can be discussed further of course):
> 
>     * `CONFIG_BT_RECV_DIRECT`
> 
>     * `CONFIG_BT_RECV_FIFO`
> 
>     * `CONFIG_BT_RECV_SYS_WORKQ`
> 
>     * `CONFIG_BT_RECV_USR_WORKQ`
> 
> 
> This way users would be able to choose what to run most of the BLE stack on, they wouldn't be forced to a single model.
> 
> We would default to `CONFIG_BT_RECV_DIRECT` so that we wouldn't need to change the system workqueue stack size by default, instead asking users to do so if they select the `CONFIG_BT_RECV_SYS_WORKQ` option

Interesting idea. If we do implement support for multiple solutions here, can we ensure that the behavior of the BT stack remains the same? e.g. already now with 2 options (RX thread vs system workqueue), the documentation states that just those two will give different behavior. If we start introducing more options, we need to ensure 1 single specific behavior, otherwise it will be a complete mess.

LingaoM commented:
I hope you can reach an agreement and try to avoid the code changes I've been making.

carlescufi commented:
> I hope you can reach an agreement and try to avoid the code changes I've been making.

We are currently deliberating, I will create a thread in Discord to continue the discussion.

In the meantime, could you please take out the `Bluetooth: Host: Fix bt_recv_thread_id not correctly assignment` commit and post it in a separate PR?

alwa-nordic commented:
> > Since there is no consensus on automation configuration and no constructive comments from anyone other than @Thalley , considering how long this PR exists, I do not intend to continue to invest in a pointless thing, so choose to close and wait @jhedberg so called 3rd option?
> 
> @LingaoM @jhedberg @Thalley We discussed this today and, in order to be more conservative, how about we change it in this way:
> 
> Change `CONFIG_BT_RECV_IS_RX_THREAD` into a `choice`: `CONFIG_BT_RECV_CONTEXT` with the following options (names can be discussed further of course):
> 
>     * `CONFIG_BT_RECV_DIRECT`
> 
>     * `CONFIG_BT_RECV_FIFO`
> 
>     * `CONFIG_BT_RECV_SYS_WORKQ`
> 
>     * (and optionally) `CONFIG_BT_RECV_USR_WORKQ`
> 
> 
> This way users would be able to choose what to run most of the BLE stack on, they wouldn't be forced to a single model.
> 
> We would default to `CONFIG_BT_RECV_DIRECT` so that we wouldn't need to change the system workqueue stack size by default, instead asking users to do so if they select the `CONFIG_BT_RECV_SYS_WORKQ` option

I agree. The names for the choices you propose are good. Adding WORKQ as a parallel implementation will let uses test it and let us gain confidence in the correctness before making it the default or mandatory option. But we should default to FIFO because it's the safest and also the same as the current default of `BT_RECV_IS_RX_THREAD=n`. +1

rugeGerritsen commented on subsys/bluetooth/host/Kconfig:
```diff
@@ -70,19 +70,48 @@ config BT_HCI_RESERVE
+	  Indicate that host will create another thread, only used as BT RX.
+
+config BT_RECV_SYS_WORKQ
+	bool "System work queue used"
```
Process low 
```suggestion
	bool "Process low priority HCI packets in the system work queue"
```

rugeGerritsen commented on subsys/bluetooth/host/Kconfig:
```diff
@@ -70,19 +70,48 @@ config BT_HCI_RESERVE
+config BT_RECV_SYS_WORKQ
+	bool "System work queue used"
+	help
+	  System work queue used as bluetooth receive thread.
+
+	  Note: Refer to BT_RX_STACK_SIZE for the recommended minimum
+	  stack space of the system work queue.
```
```suggestion
	  When this option is selected, the host will process incoming low priority HCI packets
	  in the system work queue. The HCI driver shall not call bt_recv_prio().
	  High priority HCI packets will processed in the context of the caller of bt_recv().
	  The application needs to ensure the system workqueue stack size is large enough, 
	  refer to BT_RX_STACK_SIZE for the recommended minimum.

	  Note: When this option is used, other users of the system work queue will influence the 
	  latency of incoming Bluetooth events.
```

rugeGerritsen commented on subsys/bluetooth/host/Kconfig:
```diff
@@ -70,19 +70,48 @@ config BT_HCI_RESERVE
+	  stack space of the system work queue.
+
+config BT_RECV_USR_WORKQ
+	bool "User work queue used"
```
```suggestion
	bool "Process low priority HCI packets in a user work queue"
```

carlescufi replied:
Remove this one for now, please

rugeGerritsen commented on subsys/bluetooth/host/Kconfig:
```diff
@@ -70,19 +70,48 @@ config BT_HCI_RESERVE
+config BT_RECV_USR_WORKQ
+	bool "User work queue used"
+	help
+	  User work queue used as bluetooth receive thread.
+
+	  Note: Refer to BT_RX_STACK_SIZE for the recommended minimum
+	  stack space of the system work queue.
```
```suggestion
	  This option is similar to BT_RECV_SYS_WORKQ, but with the advantage of
	  making Bluetooth processing independent of the system work queue.
	  The application is responsible for calling bt_rx_usr_workq_init() before bt_enable().
```

rugeGerritsen commented on include/bluetooth/bluetooth.h:
```diff
@@ -116,6 +116,17 @@ struct bt_le_ext_adv_cb {
+/**
+ * @brief Registered user work queue.
+ *
+ * when @kconfig{CONFIG_BT_RECV_USR_WORKQ} is enabled, the application must
+ * enable user work queue and call this API to register before Bluetooth APIs
+ * can be used.
```
```suggestion
 * When @kconfig{CONFIG_BT_RECV_USR_WORKQ} is enabled, the application must
 * set which user work queue to be used for processing incoming HCI packets. 
 * This API must be called before bt_enabled().
```

rugeGerritsen commented on include/bluetooth/gatt.h:
```diff
@@ -1216,7 +1216,8 @@ struct bt_gatt_exchange_params {
  *  @note Shall only be used once per connection.
  *
  *  The Response comes in callback @p params->func. The callback is run from
- *  the BT RX thread. @p params must remain valid until start of callback.
+ *  the BT RX thread, depends on `config BT_RECV_CONTEXT` selection.
```
I prefer not mentioning the BT RX thread here as it may not be a thread.
```suggestion
 *  The Response comes in callback @p params->func. The callback is run from
 *  the context specified by 'config BT_RECV_CONTEXT'.
 *  @p params must remain valid until start of callback.
```
Similar in other occurrences.

rugeGerritsen commented on subsys/bluetooth/host/Kconfig:
```diff
@@ -70,19 +70,48 @@ config BT_HCI_RESERVE
 config BT_RX_STACK_SIZE
 	int "Size of the receiving thread stack"
-	depends on BT_HCI_HOST || BT_RECV_IS_RX_THREAD
+	depends on BT_HCI_HOST || BT_RECV_DIRECT
```
I maybe be misunderstanding something here. 
Why is it not 
```suggestion
	depends on BT_HCI_HOST || BT_RECV_THREAD
```
?

I thought this is the stack size to be used when there is a thread in the host. Similar for BT_RX_PRIO

rugeGerritsen commented on tests/bluetooth/hci_prop_evt/prj.conf:
```diff
@@ -4,7 +4,7 @@ CONFIG_ZTEST=y
 CONFIG_BT_CTLR=n
 CONFIG_BT_NO_DRIVER=y
-CONFIG_BT_RECV_IS_RX_THREAD=y
+CONFIG_BT_RECV_DIRECT=y
```
Why do we need to set `CONFIG_BT_RECV_DIRECT=y` explicitly? Similar in `host_long_adv_recv`

rugeGerritsen reviewed (changes requested):
I would like to document the choices better. For example: what are the benefits and drawbacks for the different options? I've provided some suggested improvements to the documentation of the Kconfig options. Corresponding changes should be put in the documentation of `bt_recv()` and `bt_recv_prio()` as well.

My assumption is that when not using BT_RECV_DIRECT, `bt_recv()` needs to be called from a cooperative thread to avoid race conditions when running the RX thread/work queue. This needs to be documented. Please correct me if I'm wrong. 

I believe this conceptual documentation belongs in https://github.com/zephyrproject-rtos/zephyr/blob/main/doc/guides/bluetooth/bluetooth-arch.rst

carlescufi commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -61,9 +61,16 @@
+#elif defined(CONFIG_BT_RECV_SYS_WORKQ)
+static void rx_work_handler(struct k_work *work);
+static K_WORK_DEFINE(rx_work, rx_work_handler);
+#elif defined(CONFIG_BT_RECV_USR_WORKQ)
```
Please remove this

carlescufi commented:
@LingaoM thank you for reworking this, we are close to coming to an agreement.

Could you please remove the `CONFIG_BT_RECV_USR_WORKQ` handling for now? Everything: the Kconfig choice and all the code referencing it.

Thanks!

carlescufi commented:
@LingaoM could you please ensure in `subsys/bluetooth/common/dummy.c`:

```
#if defined(CONFIG_BT_RECV_SYS_WORKQ )
BUILD_ASSERT(CONFIG_BT_HCI_TX_PRIO < CONFIG_SYSTEM_WORKQUEUE_PRIORITY );
#endif
```

carlescufi commented on subsys/bluetooth/host/Kconfig:
```diff
@@ -70,19 +70,48 @@ config BT_HCI_RESERVE
+
+choice BT_RECV_CONTEXT
+	prompt "BT RX Thread Selection"
+	default BT_RECV_DIRECT if BT_LL_SW_SPLIT || BT_H4
+	default BT_RECV_THREAD
```
```suggestion
	default BT_RECV_THREAD
```

And then `select BT_RECV_DIRECT` in the BT_LL_SW_SPLIT and BT_H4 Kconfigs

LingaoM replied:
warning: the choice symbol BT_RECV_BLOCKING (defined at subsys/bluetooth/host/Kconfig:83) is selected by the following symbols, but select/imply has no effect on choice symbols
 - BT_LL_SW_SPLIT (defined at subsys/bluetooth/controller/Kconfig:97)

LingaoM replied:
We cannot select the config of a choice, so we can only do so.

jori-nordic commented on subsys/bluetooth/host/Kconfig:
```diff
@@ -70,19 +70,48 @@ config BT_HCI_RESERVE
+	help
+	  Select the Bluetooth RX Thread.
+
+config BT_RECV_DIRECT
```
Could we rename this to `BT_RECV_BLOCKING` and `BT_RECV_THREAD` to `BT_RECV_NONBLOCKING`, it's more obvious what's happening then.

jori-nordic reviewed (commented):
I think we could simplify a bit, by getting rid of the BT RX thread entirely, transforming it into a bluetooth-only WQ instead. I don't think it would change the behavior re:priorities if we use the same prio for the new bt workqueue. 

Then the code path when handling events/data would just depend on whether bt_recv() is blocking or non-blocking (see my other comment about renaming the kconfig options).

carlescufi commented:
@LingaoM  we reached a decision in [this Discord thread](https://discord.com/channels/720317445772017664/958293471674908672/959074904454606918):

The choice should be:

- `BT_RECV_BLOCKING`  (= `CONFIG_BT_RECV_IS_RX_THREAD =y`)
- `BT_RECV_WORKQ_BT` (= `CONFIG_BT_RECV_IS_RX_THREAD =n`, using a Bluetooth-specific workqueue)
- `BT_RECV_WORKQ_SYS` (= `CONFIG_BT_RECV_IS_RX_THREAD =n`, using the system workqueue)

LingaoM commented:
> @LingaoM we reached a decision in [this Discord thread](https://discord.com/channels/720317445772017664/958293471674908672/959074904454606918):
> 
> The choice should be:
> 
> * `BT_RECV_BLOCKING`  (= `CONFIG_BT_RECV_IS_RX_THREAD =y`)
> * `BT_RECV_WORKQ_BT` (= `CONFIG_BT_RECV_IS_RX_THREAD =n`, using a Bluetooth-specific workqueue)
> * `BT_RECV_WORKQ_SYS` (= `CONFIG_BT_RECV_IS_RX_THREAD =n`, using the system workqueue)

I don't think it's ideal to create a user workq. It needs to select userspace and depends on other config. Maybe it's more appropriate for us to directly use the original separate thread?

carlescufi commented:
> I don't think it's ideal to create a user workq. It needs to select userspace and depends on other config. Maybe it's more appropriate for us to directly use the original separate thread?

BT_RECV_BLOCKING:  process everything directly in `bt_recv()`
BT_RECV_WORKQ_SYS: Defer processing to the system workqueue

BT_RECV_WORKQ_BT: Defer processing to a new workqueue, created with:

in `hci_core.c`:
```
#ifdef BT_RECV_WORKQ_BT
static K_KERNEL_STACK_DEFINE(bt_work_q_stack,
			     CONFIG_BT_WORKQUEUE_STACK_SIZE);
struct k_work_q bt_work_q;

	k_work_queue_start(&bt_work_q,
			    bt_work_q_stack,
			    K_KERNEL_STACK_SIZEOF(bt_work_q_stack),
			    CONFIG_BT_WORKQUEUE_PRIORITY, &cfg);
#endif
```

and remove all the code that currently creates:
```
REMOVE:
static struct k_thread rx_thread_data;
static K_KERNEL_STACK_DEFINE(rx_thread_stack, CONFIG_BT_RX_STACK_SIZE);
static void hci_rx_thread(void)
{
}
```

Thalley commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -3413,6 +3417,22 @@ void hci_event_prio(struct net_buf *buf)
 	}
 }
 
+#if !defined(CONFIG_BT_RECV_BLOCKING)
+static void rx_queue_put(struct net_buf *buf)
+{
+	net_buf_put(&bt_dev.rx_queue, buf);
+
+#if defined(CONFIG_BT_RECV_WORKQ_SYS)
+	int err = k_work_submit(&rx_work);
+#elif defined(CONFIG_BT_RECV_WORKQ_BT)
+	int err = k_work_submit_to_queue(&bt_workq, &rx_work);
+#endif /* CONFIG_BT_RECV_WORKQ_SYS */
+	if (err < 0) {
+		BT_ERR("Could not submit rx_work: %d", err);
+	}
+}
+#endif /* !CONFIG_BT_RECV_BLOCKING */
```
This one is a bit fun, because the function itself is guarded with `!defined(CONFIG_BT_RECV_BLOCKING)` and the individual work items are guarded with `CONFIG_BT_RECV_WORKQ_SYS` and `CONFIG_BT_RECV_WORKQ_BT`. 

So in the case that we add another WORKQ option, this function will start compile properly, but won't work as intended. 

I would suggest to guard the function with `#if defined(CONFIG_BT_RECV_WORKQ_SYS) || defined(CONFIG_BT_RECV_WORKQ_BT)` instead, to ensure that it isn't called without support, and also commonlize a bit more:

```suggestion
#if defined(CONFIG_BT_RECV_WORKQ_SYS) || defined(CONFIG_BT_RECV_WORKQ_BT)
static void rx_queue_put(struct net_buf *buf)
{
	net_buf_put(&bt_dev.rx_queue, buf);

#if defined(CONFIG_BT_RECV_WORKQ_SYS)
	const int err = k_work_submit(&rx_work);
#elif defined(CONFIG_BT_RECV_WORKQ_BT)
	const int err = k_work_submit_to_queue(&bt_workq, &rx_work);
#endif /* CONFIG_BT_RECV_WORKQ_SYS */

	if (err < 0) {
		BT_ERR("Could not submit rx_work: %d", err);
	}
}
#endif /* CONFIG_BT_RECV_WORKQ_SYS || CONFIG_BT_RECV_WORKQ_BT */
```

LingaoM commented on include/bluetooth/gatt.h:
```diff
@@ -1216,7 +1216,8 @@ struct bt_gatt_exchange_params {
  *
  *  The Response comes in callback @p params->func. The callback is run from
- *  the BT RX thread. @p params must remain valid until start of callback.
+ *  the context specified by 'config BT_RECV_CONTEXT'.
```
Error: See https://docs.zephyrproject.org/latest/guides/kconfig/index.html for more details.

Found references to undefined Kconfig symbols. If any of these are false
positives, then add them to UNDEF_KCONFIG_WHITELIST in check_compliance.py in the ci-tools repo.

If the reference is for a comment like /* CONFIG_FOO_* */ (or
/* CONFIG_FOO_*_... */), then please use exactly that form (with the '*'). The
CI check knows not to flag it.

More generally, a reference followed by $, @, {, *, or ## will never be
flagged.

CONFIG_BT_RECV_CONTEXT                     include/bluetooth/gatt.h:1219, include/bluetooth/gatt.h:1461, include/bluetooth/gatt.h:1509, include/bluetooth/gatt.h:1699
Error: Process completed with exit code 1.

LingaoM replied:
Maybe we can't use it like this. For `choice` config

Thalley replied:
That's a bit odd, but yeah maybe we can't use `@kconfig{...}` on choices. I guess we'll just have to omit the `@kconfig`

Thalley commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -61,10 +61,14 @@
 /* Stacks for the threads */
-#if !defined(CONFIG_BT_RECV_IS_RX_THREAD)
-static struct k_thread rx_thread_data;
+#if !defined(CONFIG_BT_RECV_BLOCKING)
```
We could consider doing 
```suggestion
#if defined(CONFIG_BT_RECV_WORKQ_SYS) || defined(CONFIG_BT_RECV_WORKQ_BT)
```
here as well, but I'll leave that up to you. It is OK as is.

Thalley reviewed (dismissed):
LGTM

LingaoM commented:
@rugeGerritsen @carlescufi Please take a look.

LingaoM commented:
@rugeGerritsen @carlescufi Please take a look.

carlescufi reviewed (dismissed):
LGTM, nice work!

carlescufi commented:
Please DNM until @cvinayak and @jhedberg have approved it.

jori-nordic commented on subsys/bluetooth/controller/Kconfig:
```diff
@@ -96,7 +96,6 @@ choice BT_LL_CHOICE
 
 config BT_LL_SW_SPLIT
 	bool "Software-based BLE Link Layer"
-	select BT_RECV_IS_RX_THREAD
```
don't we want to keep this @cvinayak ? I don't see where the option is selected otherwise.

LingaoM replied:
https://github.com/zephyrproject-rtos/zephyr/pull/40273#discussion_r838438179

jori-nordic replied:
fair enough, approved :)

carlescufi commented:
@LingaoM please rebase

LingaoM commented:
> @LingaoM please rebase

This is DNM label for this？

carlescufi commented:
> > @LingaoM please rebase
> 
> This is DNM label for this？

Waiting for @cvinayak to approve. Can you rebase @LingaoM ?
